### PR TITLE
add spdata to tools

### PR DESCRIPTION
**Pull request by bqpd**
as used to create examples in https://github.com/hoburg/gpkit-models/issues/8

### CODE REVIEW

**Comment by bqpd** [comment]
@whoburg, ready for review

**Comment by bqpd** [comment]
retest this please

**Comment by galbramc** [comment]
This adds h5py as another dependency. Is that intentional? I need to update the scripts to install it if so.

**Comment by bqpd** [comment]
I don't think we should add it as a dependency, no. This was a useful tool
that may be used by future tests, but until that point we're okay.

**Comment by galbramc** [comment]
Hmm, well I need to add it to the jenkins script, or you need to suppress the pylint warning that h5py can't be imported. @whoburg, which solution would you prefer?

File: gpkit/tools/spdata.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    Usage
+    -----
+
+    ``spdata = SPData(m)``
+    ``spdata.save('example_sp.h5')``
+    """
+    def __init__(self, model):
+        # pylint:disable=super-init-not-called
+        if not hasattr(model, "solution"):
+            print "You need to solve the model first."
```
Comment (whoburg) [inline gpkit/tools/spdata.py]: raise a `RuntimeWarning` here?

File: gpkit/tools/spdata.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+        self.signomials = [model.cost]
+        for constraint in model.flat(constraintsets=False):
+            if isinstance(constraint, (SignomialInequality,
+                                       PosynomialInequality)):
+                self.signomials.extend(constraint.unsubbed)
```
Comment (whoburg) [inline gpkit/tools/spdata.py]: else... print a warning / raise an Exception?

File: gpkit/tools/spdata.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+        # k [j]: number of monomials (columns of F) present in each constraint
+        k = [len(p.cs) for p in self.signomials]
+        # p_idxs [i]: posynomial index of each monomial
+        p_idxs = []
+        for i, p_len in enumerate(k):
+            p_idxs += [i]*p_len
+        self.p_idxs = np.array(p_idxs)
+        # A [i, v]: sparse matrix of variable's powers in each monomial
+        self.A, _ = genA(self.exps, self.varlocs)
```
Comment (whoburg) [inline gpkit/tools/spdata.py]: inline TODO suggesting that we get these from self.A, etc once NomialData has those properties?

File: gpkit/tools/spdata.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    def __repr__(self):
+        return str(self.signomials)
+
+    def save(self, filename):
+        "Save spdata to an h5 file."
+        try:
+            import h5py  # pylint:disable=import-error
+        except ImportError as ie:
+            print "SPData.save requires the h5py library."
+            raise ie
```
Comment (whoburg) [inline gpkit/tools/spdata.py]: could assert that if `filename` has an extension, it isn't something other than `.h5`

File: gpkit/tools/spdata.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        self.A, _ = genA(self.exps, self.varlocs)
+
+        self.varsols = np.array([mag(model.solution(var))
+                                 for var in self.varlocs])
+        self.varnames = np.array([str(var) for var in self.varlocs])
+
+    def __repr__(self):
+        return str(self.signomials)
+
+    def save(self, filename):
```
Comment (whoburg) [inline gpkit/tools/spdata.py]: maybe this should be the `save` method of `NomialData`, to better set ourselves up for doing the same thing with text, csv, etc?

File: gpkit/tools/spdata.py (hunk 1 above)
Comment (bqpd) [inline gpkit/tools/spdata.py]: Good point. Probably a `ValueError`

File: gpkit/tools/spdata.py (hunk 2 above)
Comment (bqpd) [inline gpkit/tools/spdata.py]: Yup, should probably print a warning.

File: gpkit/tools/spdata.py (hunk 4 above)
Comment (bqpd) [inline gpkit/tools/spdata.py]: I think filename is a reasonable thing to expect the user to handle...

**Pull request merged**